### PR TITLE
fix: better mismatch errors

### PR DESCRIPTION
**Pull request by erights**
For the pattern language to actually be used a lot, it needs to emit better errors diagnosing a pattern match failure.

TODO After this is merged, merge https://github.com/Agoric/documentation/pull/695

### CODE REVIEW

File: packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js [hunk 1]
```diff
@@ -299,13 +299,13 @@ test(`zoeHelper with zcf - fit proposal patterns`, async t => {
...
   t.throws(() => fit(proposal, harden([])), {
-    message: /.* - Must be equivalent to: \[\]/,
+    message: /.* - Must be: \[\]/,
   });
   t.throws(
     () => fit(proposal, M.split({ want: { C: M.any() } })),
     {
       message:
-        /Must have same property names as record pattern: {"C":"\[match:any\]"}/,
+        /want: {"A":{"brand":"\[Alleged: moola brand\]","value":"\[20n\]"}} - Must have missing properties \["C"\]/,
```
Comment (Chris-Hibbert) [inline packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js]: can you get rid of the backslashes and/or square brackets in the property name?

File: packages/store/src/patterns/patternMatchers.js [hunk 2]
```diff
@@ -295,12 +308,32 @@ const makePatternKit = () => {
...
+            return check(
+              false,
+              X`${specimen} - Must have missing properties ${q(missing)}`,
+            );
+          }
+          const passNameSet = new Set(pattNames);
+          const unexpected = specNames.filter(name => !passNameSet.has(name));
+          assert(
+            unexpected.length >= 1,
+            X`Internal: must have either missing or extra: ${q(
```
Comment (Chris-Hibbert) [inline packages/store/src/patterns/patternMatchers.js:323]: This doesn't show up in any tests. Should be easy to provoke, right?

File: packages/store/src/patterns/patternMatchers.js [hunk 3]
```diff
@@ -255,15 +255,28 @@ const makePatternKit = () => {
...
+  /**
+   * @param {Passable} specimen
+   * @param {Pattern} pattern
+   * @param {Checker=} check
+   * @param {string|number} [label]
+   * @returns {boolean}
+   */
+  const checkMatches = (specimen, pattern, check = x => x, label = undefined) =>
+    // eslint-disable-next-line no-use-before-define
+    applyLabelingError(checkMatchesInternal, [specimen, pattern, check], label);
```
Comment (Chris-Hibbert) [inline packages/store/src/patterns/patternMatchers.js:267]: I can't find any tests that follow the error path, so the label isn't detectably  checked.

File: packages/store/src/patterns/patternMatchers.js [hunk 4]
```diff
@@ -976,12 +1034,8 @@ const makePatternKit = () => {
...
       harden(newBase);
       return (
-        checkMatches(specB, newBase, check) &&
-        (rest === undefined ||
-          check(
-            matches(specR, rest),
-            X`Remainder ${specR} - Must match ${rest}`,
-          ))
+        checkMatches(specB, newBase, check, 'partial-base') &&
+        (rest === undefined || checkMatches(specR, rest, check, 'partial-rest'))
```
Comment (Chris-Hibbert) [inline packages/store/src/patterns/patternMatchers.js]: Is there something we could expand *partial-base* and *partial-rest* to that would be more meaningful in context? I'm not precisely sure what the semantics is, so I don't have a suggestion, but these don't tell me much when looking at the relevant errors.

**Review by Chris-Hibbert — CHANGES_REQUESTED** [review]
This is an improvement. I have a few suggestions for further improvements, and some requests for additional tests.

File: packages/store/src/patterns/patternMatchers.js (hunk 2 above)
Comment (erights) [reply packages/store/src/patterns/patternMatchers.js:323]: I label an error `Internal:` when it means that it should not be able to happen, so if it does, something is seriously screwed up. We should also hook these up to more severe termination (like of the vat), but there's not an easy way to do that here.

File: packages/store/src/patterns/patternMatchers.js (hunk 4 above)
Comment (erights) [reply packages/store/src/patterns/patternMatchers.js]: Agreed. And that was my second try on terminology. Maybe brainstorm sometime soon?

File: packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js (hunk 1 above)
Comment (erights) [reply packages/zoe/test/unitTests/zcf/test-zoeHelpersWZcf.js]: By changing many of these from regexps to literal strings, I was able to get rid of many backslashes. More readable, thanks!

Done. Resolving.

File: packages/store/src/patterns/patternMatchers.js (hunk 2 above)
Comment (erights) [reply packages/store/src/patterns/patternMatchers.js:323]: I don't think there's anything to do here, so Resolving. But feel free to unresolve and discuss.

File: packages/store/src/patterns/patternMatchers.js (hunk 3 above)
Comment (erights) [reply packages/store/src/patterns/patternMatchers.js:267]: For example, in test-inputValidation we see the error message
```
        'displayInfo: partial-base: decimalPlaces: "hello" - Must be >= -100',
```
Every level of `"<label>: "` before the `"hello"` was added on the error path by this mechanism. Resolving.

**Review by Chris-Hibbert — APPROVED** [review]
This LGTM. Ping me when you want to chat about terminology.

File: packages/store/src/patterns/patternMatchers.js (hunk 4 above)
Comment (erights) [reply packages/store/src/patterns/patternMatchers.js]: changed to required-parts optional-parts rest-parts. Much better!

**Pull request merged**